multiapps-controller-persistence/src/main/java/org/cloudfoundry/multiapps/controller/persistence/services/GcpObjectStoreFileStorage.java

```diff
@@ -0,0 +1,234 @@
+    public List<FileEntry> getFileEntriesWithoutContent(List<FileEntry> fileEntries) throws FileStorageException {
+        Set<String> existingFiles = getAllEntries().stream()
+                                                   .map(Blob::getName)
+                                                   .collect(Collectors.toSet());
+        return fileEntries.stream()
+                          .filter(fileEntry -> !existingFiles.contains(fileEntry.getId()))
+                          .toList();
+    }
+
+    @Override
+    public void deleteFile(String id, String space) {
+        storage.delete(bucketName, id);
+    }
+
+    @Override
+    public void deleteFilesBySpaceIds(List<String> spaceIds) {
+        removeBlobsByFilter(blob -> ObjectStoreUtil.filterBySpaceIds(blob.getMetadata(), spaceIds));
+    }
+
+    @Override
+    public void deleteFilesBySpaceAndNamespace(String space, String namespace) {
+        removeBlobsByFilter(blob -> ObjectStoreUtil.filterBySpaceAndNamespace(blob.getMetadata(), space, namespace));
+    }
+
+    @Override
+    public int deleteFilesModifiedBefore(LocalDateTime modificationTime) {
+        return removeBlobsByFilter(
+            blob -> ObjectStoreUtil.filterByModificationTime(blob.getMetadata(), blob.getName(), modificationTime));
+    }
+
+    private InputStream getBlobPayloadWithOffset(FileEntry fileEntry, long startOffset, long endOffset)
+        throws FileStorageException {
+        try {
+            return getBlobWithRetriesWithOffset(fileEntry, 3, startOffset, endOffset);
+        } catch (IOException e) {
+            throw new FileStorageException(e);
+        }
+    }
+
+    @Override
+    public <T> T processFileContent(String space, String id,
+                                    FileContentProcessor<T> fileContentProcessor) throws FileStorageException {
+        FileEntry fileEntry = ObjectStoreUtil.createFileEntry(space, id);
+        try (InputStream inputStream = openBlobStreamWithRetries(fileEntry, 3)) {
+            if (inputStream == null) {
+                throw new FileStorageException(
+                    MessageFormat.format(Messages.FILE_WITH_ID_AND_SPACE_DOES_NOT_EXIST,
+                                         fileEntry.getId(), fileEntry.getSpace()));
+            }
+            return fileContentProcessor.process(inputStream);
+        } catch (Exception e) {
+            throw new FileStorageException(e);
+        }
+    }
+
+    private InputStream openBlobStreamWithRetries(FileEntry fileEntry, int maxAttempts) {
+        Blob blob = getBlobWithRetries(fileEntry, maxAttempts);
+        if (blob == null) {
+            return null;
+        }
+        return Channels.newInputStream(blob.reader());
+    }
+
+    private Blob getBlobWithRetries(FileEntry fileEntry, int retries) {
+        for (int i = 1; i <= retries; i++) {
+            Blob blob = storage.get(bucketName, fileEntry.getId());
+            if (blob != null) {
+                return blob;
+            }
```

Comment:
does this retry actually work? the blob will be null only when it does not exist. don't we need to retry in case of exception?